planner/core/task.go

```diff
@@ -872,54 +872,59 @@ func (sel *PhysicalSelection) attach2Task(tasks ...task) task {
 	return t
 }
 
-func (p *basePhysicalAgg) newPartialAggregate(copToFlash bool) (partial, final PhysicalPlan) {
-	// Check if this aggregation can push down.
-	sc := p.ctx.GetSessionVars().StmtCtx
-	client := p.ctx.GetClient()
-	for _, aggFunc := range p.AggFuncs {
+// CheckAggCanPushCop checks whether the aggFuncs with groupByItems can
+// be pushed down to coprocessor.
+func CheckAggCanPushCop(aggFuncs []*aggregation.AggFuncDesc, groupByItems []expression.Expression, sctx sessionctx.Context, copToFlash bool) bool {
+	sc := sctx.GetSessionVars().StmtCtx
+	client := sctx.GetClient()
+	for _, aggFunc := range aggFuncs {
 		if copToFlash {
 			if !aggregation.CheckAggPushFlash(aggFunc) {
-				return nil, p.self
+				return false
 			}
-			if _, remain := expression.CheckExprPushFlash(append(aggFunc.Args, p.GroupByItems...)); len(remain) > 0 {
-				return nil, p.self
+			if _, remain := expression.CheckExprPushFlash(append(aggFunc.Args, groupByItems...)); len(remain) > 0 {
+				return false
 			}
 		}
 		pb := aggregation.AggFuncToPBExpr(sc, client, aggFunc)
 		if pb == nil {
-			return nil, p.self
+			return false
 		}
 	}
-	_, _, remained := expression.ExpressionsToPB(sc, p.GroupByItems, client)
+	_, _, remained := expression.ExpressionsToPB(sc, groupByItems, client)
 	if len(remained) > 0 {
-		return nil, p.self
+		return false
 	}
+	return true
+}
 
-	finalSchema := p.schema
-	partialSchema := expression.NewSchema()
-	p.schema = partialSchema
-	partialAgg := p.self
-
+// SplitAggToPartial1 splits either LogicalAggregation or PhysicalAggregation to finalAgg and partial1Agg.
+func SplitAggToPartial1(
+	aggFuncs []*aggregation.AggFuncDesc,
+	groupByItems []expression.Expression,
+	finalSchema *expression.Schema,
+	sctx sessionctx.Context) (finalAggFuncs []*aggregation.AggFuncDesc, finalGbyItems []expression.Expression, partialSchema *expression.Schema) {
```

Comment:
This function name is so confusing, given an aggregation and schema, the function outputs final aggregation and partial schema, while the function name is `SplitAggToPartial1` ...

Reply:
How about `BuildFinalModeAggregation`?